etl/steps/data/garden/energy/2022-12-12/electricity_mix.py

```diff
@@ -0,0 +1,324 @@
+
+"""
+
+from typing import Dict, List
+
+import pandas as pd
+from owid import catalog
+from owid.datautils.dataframes import combine_two_overlapping_dataframes
+from shared import CURRENT_DIR, add_population
+
+from etl.paths import DATA_DIR
+
+# Details for dataset to export.
+DATASET_SHORT_NAME = "electricity_mix"
+METADATA_PATH = CURRENT_DIR / f"{DATASET_SHORT_NAME}.meta.yml"
+# Details for datasets to import.
+BP_DATASET_PATH = DATA_DIR / "garden/bp/2022-07-14/statistical_review"
+EMBER_DATASET_PATH = DATA_DIR / "garden/ember/2022-08-01/combined_electricity_review"
+
+# Conversion factors.
+# Terawatt-hours to kilowatt-hours.
+TWH_TO_KWH = 1e9
+# Megatonnes to grams.
+MT_TO_G = 1e12
+
+
+def process_bp_data(table_bp: catalog.Table) -> pd.DataFrame:
+    """Load necessary columns from BP's Statistical Review dataset, and create some new variables (e.g. electricity
+    generation from fossil fuels).
+
+    Parameters
+    ----------
+    table_bp : catalog.Table
+        BP's Statistical Review (already processed, with harmonized countries and region aggregates).
+
+    Returns
+    -------
+    df_bp : pd.DataFrame
+        Processed BP data.
+
+    """
+    # Columns to load from BP dataset.
+    columns = {
+        "electricity_generation": "total_generation__twh",
+        "primary_energy_consumption__twh": "primary_energy_consumption__twh",
+        "hydro_generation__twh": "hydro_generation__twh",
+        "nuclear_generation__twh": "nuclear_generation__twh",
+        "solar_generation__twh": "solar_generation__twh",
+        "wind_generation__twh": "wind_generation__twh",
+        "geo_biomass_other__twh": "other_renewables_including_bioenergy_generation__twh",
+        "elec_gen_from_oil": "oil_generation__twh",
+        "elec_gen_from_coal": "coal_generation__twh",
+        "elec_gen_from_gas": "gas_generation__twh",
+    }
+    table_bp = table_bp[list(columns)].rename(columns=columns, errors="raise")
+    # New columns to be created by summing other columns.
+    aggregates: Dict[str, List[str]] = {
+        "fossil_generation__twh": [
+            "oil_generation__twh",
+            "coal_generation__twh",
+            "gas_generation__twh",
+        ],
+        "renewable_generation__twh": [
+            "hydro_generation__twh",
+            "solar_generation__twh",
+            "wind_generation__twh",
+            "other_renewables_including_bioenergy_generation__twh",
+        ],
+        "low_carbon_generation__twh": [
+            "renewable_generation__twh",
+            "nuclear_generation__twh",
+        ],
+    }
+
+    # Create new columns, by adding up other columns (and allowing for only one nan in each sum).
+    for new_column in aggregates:
+        table_bp[new_column] = pd.DataFrame(table_bp)[aggregates[new_column]].sum(
+            axis=1, min_count=len(aggregates[new_column]) - 1
+        )
+
+    # Prepare data in a dataframe with a dummy index.
+    df_bp = pd.DataFrame(table_bp).reset_index()
```

Comment:
If `df_bp` has already been defined (see my previous comment above), then you'd simply `df_bp = df_bp.reset_index()`